tests/Unit/Domain/CoordinateMaps/Test_CylindricalEndcap.cpp

```diff
@@ -0,0 +1,218 @@
+    // invertible is that proj_center cannot lie between sphere_one
+    // and the center of cone_regular.
+
+    // The second cone we will construct we will call cone_opening.
+    // It will have some maximum opening angle that we choose freely.
+    const double max_opening_angle = M_PI / 3.0;
+    // max_opening_angle determines the maximum possible x-coordinate
+    // of proj_center.
+    const double max_proj_center_z =
+        z_plane -
+        radius_one * sqrt(1.0 - square(cos_theta)) / tan(max_opening_angle);
+    // Cone_opening has apex
+    // (center_one[0], center_one[1], max_proj_center_z). Note that if
+    // max_opening_angle > pi/4 and if z_plane > center_one[2] then
+    // the apex of cone_opening is inside sphere_one.
+
+    // Now that we have two cones, cone_regular and cone_opening.  We
+    // will choose proj_center so that it lies inside of both cones,
+    // and that it lies inside of sphere_two.
+    const double cone_regular_opening_angle = asin(cos_theta);
+
+    if (max_opening_angle < cone_regular_opening_angle) {
+      // The easier case.  We need to worry only about max_opening_angle.
+
+      // Choose some smaller cone inside of cone_opening.
+      // We will place proj_center on this smaller cone.
+      const double beta = unit_dis(gen) * max_opening_angle;
+
+      // Choose an azimuthal coordinate for the point on the cone.
+      const double phi = angle_dis(gen);
+
+      // Choose a radius for the point on the cone that will be
+      // proj_center. The point should be inside sphere_two.  So
+      // determine the radius at which the point intersects
+      // sphere_two. This radius is the solution of a quadratic equation
+      // ax^2 + bx + c = 0
+      const double a = 1.0;
+      const double b =
+          2.0 * (-(max_proj_center_z - center_two[2]) * cos(beta) +
+                 (center_one[0] - center_two[0]) * sin(beta) * cos(phi) +
+                 (center_one[1] - center_two[1]) * sin(beta) * sin(phi));
+      const double c = square(max_proj_center_z - center_two[2]) +
+                       square(center_one[0] - center_two[0]) +
+                       square(center_one[1] - center_two[1]) -
+                       square(radius_two);
+      ASSERT(c < 0.0,
+             "max_proj_center_z is too negative. The apex "
+             "is not inside sphere_two");
+      // Should be two real roots. Choose the positive one.
+      const double r_max = positive_root(a, b, c);
+
+      const double proj_radius = unit_dis(gen) * r_max;
+      return std::array<double, 3>{
+          center_one[0] + proj_radius * sin(beta) * cos(phi),
+          center_one[1] + proj_radius * sin(beta) * sin(phi),
+          max_proj_center_z - proj_radius * cos(beta)};
+    } else {
+      // The more difficult case.
+
+      // We may need to try several values of alpha.
+      // When one works, exit.
+      while (true) {
+        // Choose some smaller cone inside of cone_regular.
```

Comment:
Might you occasionally get time outs if it just picks bad points too often? How likely is that to happen? How many tries on average should this take to succeed?

Reply:
I'm adding a comment saying that failure is rare and on average it should succeed on the first try.  But this particular test will change in #2587 in a way that makes failure even more rare (and easier to compute a probability for).